src/cryptonote_core/cryptonote_core.h

```diff
@@ -1118,9 +1109,7 @@ namespace cryptonote
 
      std::atomic_flag m_checkpoints_updating; //!< set if checkpoints are currently updating to avoid multiple threads attempting to update at once
 
-     bool m_service_node;
-     crypto::secret_key m_service_node_key;
-     crypto::public_key m_service_node_pubkey;
+     std::shared_ptr<service_node_keys> m_service_node_keys;
```

Comment:
Yes: the null pointer captures "not a service node".

Reply:
It's also shared with `service_node_list`.

Reply:
Also suggests the usual things you associate with smart pointers like non-trivial lifetimes, ownership complexity. But if it exists for the entirety of the program then there are no lifetimes or ownership you have to consider. I'd encode the bool in the struct and operator bool if you want the syntactic niceness, or boost optional or raw ptr it.

Its a heavy handed approach for taking advantage of nullability.

Reply:
Right, I mispoke, the sharing with `service_node_list` was the more important reason.

Reply:
As for heaviness: shared_ptr deferences are extremely light; the only overhead with it compared to a raw pointer is when you creating a copy (because of needing an atomic increment) or destroying once, but neither or those are done here (except for the initial setup and the final destruction).

Reply:
I understand shared_ptrs alleviates the cost by using atomics. The cost is just a lot cheaper using the closer to zero cost alternatives. Considering cache friendliness, code generation (excluding boost::optional I'm guessing), lifetimes and ownership. 

I can't say I particularly agree though- it's true we only really take a hit on startup and shutdown and that we're not a real time program and accesses to the keys are infrequent enough.

Reply:
> I understand shared_ptrs alleviates the cost by using atomics

No: it uses atomics only on copying/creation/destruction.  There is *no* atomic operation required for dereferencing; a smart pointer deference it is literally optimized to nothing more than a pointer dereference.

Reply:
Right and there are copies in the codebase as it is right now. In quorum_cop when you receive blocks when this could just be free. It is for sharing, but you're sharing a read only object.

Reply:
Anyway, there's a push to move forward with the 5.0.0 release. So gonna merge this in